src/saveload/marker.rs

```diff
@@ -0,0 +1,209 @@
+/// If serialization strategy needs to set marker to some entity it should use
+/// new marker allocated for `Marker::Allocator`.
+///
+/// # Example
+///
+/// ```rust,no_run
+/// extern crate specs;
+/// #[macro_use] extern crate serde_derive;
+/// use std::collections::HashMap;
+/// use std::ops::Range;
+/// use specs::{Component, Entity, DenseVecStorage};
+/// use specs::saveload::marker::{Marker, MarkerAllocator};
+///
+/// // Marker for entities that should be synced over network
+/// #[derive(Clone, Copy, Serialize, Deserialize)]
+/// struct NetMarker {
+///     id: u64,
+///     seq: u64,
+/// }
+///
+/// impl Component for NetMarker {
+///     type Storage = DenseVecStorage<Self>;
+/// }
+///
+/// impl Marker for NetMarker {
+///     type Identifier = u64;
+///     type Allocator = NetNode;
+///
+///     fn id(&self) -> u64 {
+///         self.id
+///     }
+///
+///     // Updates sequence id.
+///     // Entities with too old sequence id get deleted.
+///     fn update(&mut self, update: Self) {
+///         assert_eq!(self.id, update.id);
+///         self.seq = update.seq;
+///     }
+/// }
+///
+/// // Each client and server has one
+/// // Contains id range and `NetMarker -> Entity` mapping
+/// struct NetNode {
+///     range: Range<u64>,
+///     mapping: HashMap<u64, Entity>,
+/// }
+///
+/// impl MarkerAllocator<NetMarker> for NetNode {
+///     fn allocate(&mut self, entity: Entity, id: Option<u64>) -> NetMarker {
+///         let id = id.unwrap_or_else(|| {
+///             self.range.next().expect("Id range must be virtually endless")
+///         });
+///         let marker = NetMarker {
+///             id: id,
+///             seq: 0,
+///         };
+///         self.mapping.insert(id, entity);
+///         marker
+///     }
+///
+///     fn get(&self, id: u64) -> Option<Entity> {
+///         self.mapping.get(&id).cloned()
+///     }
+/// }
+///
+/// fn main() {}
+/// ```
+pub trait Marker: SerializableComponent + Copy {
+    /// Id of the marker
+    type Identifier: Copy + ::std::fmt::Debug + Eq + Hash;
+
+    /// Allocator for this `Marker`
+    type Allocator: MarkerAllocator<Self>;
+
+    /// Get this marker internal id
+    fn id(&self) -> Self::Identifier;
+
+    /// Update marker with new value.
+    /// It must preserve internal `Identifier`.
+    ///
+    /// # Panics
+    ///
+    /// Allowed to panic if `self.id() != update.id()`.
+    /// But usually implementer may ignore `update.id()` value
+    /// as deserialization algorithm ensures `id()`'s match.
+    fn update(&mut self, update: Self) {
+        ::std::mem::drop(update);
+    }
+}
+
+/// This allocator is used with `Marker` trait
+/// It provides method for allocation of `Marker`s
+/// And also should provide `Marker -> Entity` mapping
+/// `maintain` method can be implemented for cleanup and actualization
+/// # Example
+/// see docs for `Marker`
+pub trait MarkerAllocator<M: Marker>: Resource {
```

Comment:
Why not make the `Marker` an associated?

Reply:
What if you want to use one allocator for many markers?

Reply:
One allocator can implement this trait for different M